File: python/algoforge/algo_gen/generator.py
"""LLM-driven algorithm generators for S1 Phase 1.

Three public generators:
  generate_fast(brief, provider, seed=42)
  generate_balanced(brief, provider, seed=42)
  generate_max(brief, provider, seed=42, n_candidates=5, reasoning_provider=None)

Each returns (AlgoManifest, GenerationTrace).

Robust extraction: one retry on parse or schema failure with a "please fix JSON"
follow-up message. Two consecutive failures raise GenerationError.
"""
from __future__ import annotations

import json
import logging
import warnings
from dataclasses import dataclass, field
from typing import Any, Optional

from ..types import Bar
from ..backtest import BTConfig, BTResult, BacktestEngine
from ..llm.provider import LLMProvider
from ..llm.types import ChatMessage, ChatRequest
from .prompts import (
    FAST_SYSTEM,
    BALANCED_SYSTEM,
    MAX_SYSTEM,
    extract_json_block,
    render_balanced_critique,
    render_balanced_user,
    render_fast_user,
    render_max_critique,
    render_max_user,
)
from .schema import validate_manifest
from .runtime import ManifestAlgo
from .types import AlgoManifest

logger = logging.getLogger(__name__)

# ── Models / defaults ─────────────────────────────────────────────────────────

_DEFAULT_MODEL_FAST     = "llama3.1:8b"
_DEFAULT_MODEL_BALANCED = "qwen2.5:32b"
_DEFAULT_MODEL_MAX      = "o3-mini"
_DEFAULT_TEMPERATURE    = 0.5
_DEFAULT_MAX_TOKENS     = 2048
_MAX_CANDIDATES_DEFAULT = 5
_QUICK_BACKTEST_BARS    = 200


# ── GenerationError ───────────────────────────────────────────────────────────

class GenerationError(Exception):
    """Raised when the LLM fails to produce a valid manifest after retries."""


# ── GenerationTrace ───────────────────────────────────────────────────────────

@dataclass
class CandidateSummary:
    """Record of one candidate in max mode."""
    index:       int
    manifest:    Optional[dict]
    sharpe:      Optional[float]
    trades:      Optional[int]
    max_dd:      Optional[float]
    error:       Optional[str]       = None
    critiqued:   bool                = False
    final:       bool                = False


@dataclass
class CritiquePair:
    """One critique+revision pair."""
    candidate_index: int
    critique_prompt: str
    revised_manifest: Optional[dict]
    error: Optional[str] = None


@dataclass
class GenerationTrace:
    """Captures every LLM turn for auditability."""
    mode:              str
    model:             str
    reasoning_model:   Optional[str]         = None
    seed:              int                    = 42
    # Ordered list of {role, content} dicts covering every turn
    turns:             list[dict[str, str]]   = field(default_factory=list)
    # Parse failures with raw text
    parse_failures:    list[dict[str, str]]   = field(default_factory=list)
    # Max mode only
    candidates:        list[CandidateSummary] = field(default_factory=list)
    critique_pairs:    list[CritiquePair]     = field(default_factory=list)
    warnings:          list[str]              = field(default_factory=list)


# ── Internal helpers ──────────────────────────────────────────────────────────

_FIX_JSON_PROMPT = (
    "The JSON you produced is malformed or schema-invalid. "
    "Please output ONLY the corrected manifest JSON in a ```json ... ``` fence. "
    "No prose, no explanation — just the fixed JSON."
)


def _chat(
    provider: LLMProvider,
    messages: list[ChatMessage],
    model: str,
    seed: int,
    temperature: float = _DEFAULT_TEMPERATURE,
    max_tokens: int = _DEFAULT_MAX_TOKENS,
) -> str:
    """Call provider.chat and return the assistant message content."""
    req = ChatRequest(
        model=model,
        messages=messages,
        temperature=temperature,
        max_tokens=max_tokens,
        seed=seed,
    )
    resp = provider.chat(req)
    return resp.message.content


def _try_extract_and_validate(raw: str) -> tuple[dict, AlgoManifest]:
    """Extract JSON from raw text and schema-validate it.

    Returns (raw_dict, AlgoManifest) or raises ValueError.
    """
    d = extract_json_block(raw)
    manifest = validate_manifest(d)
    return d, manifest


def _record_turn(trace: GenerationTrace, role: str, content: str) -> None:
    trace.turns.append({"role": role, "content": content})


def _generate_with_retry(
    provider: LLMProvider,
    messages: list[ChatMessage],
    model: str,
    seed: int,
    trace: GenerationTrace,
    temperature: float = _DEFAULT_TEMPERATURE,
    max_tokens: int = _DEFAULT_MAX_TOKENS,
) -> tuple[dict, AlgoManifest]:
    """Call the LLM, attempt JSON extraction and schema validation.

    On first failure: send a follow-up "fix JSON" message and try once more.
    On second failure: raise GenerationError.
    """
    # First attempt
    raw = _chat(provider, messages, model, seed, temperature, max_tokens)
    _record_turn(trace, "assistant", raw)

    try:
        return _try_extract_and_validate(raw)
    except (ValueError, TypeError, KeyError) as exc:
        trace.parse_failures.append({"raw": raw, "error": str(exc)})
        logger.debug("First attempt failed (%s); retrying with fix prompt", exc)

    # Build retry conversation
    retry_messages = list(messages) + [
        ChatMessage(role="assistant", content=raw),
        ChatMessage(role="user", content=_FIX_JSON_PROMPT),
    ]
    _record_turn(trace, "user", _FIX_JSON_PROMPT)

    raw2 = _chat(provider, retry_messages, model, seed, temperature, max_tokens)
    _record_turn(trace, "assistant", raw2)

    try:
        return _try_extract_and_validate(raw2)
    except (ValueError, TypeError, KeyError) as exc2:
        trace.parse_failures.append({"raw": raw2, "error": str(exc2)})
        raise GenerationError(
            f"Failed to produce a valid manifest after 2 attempts. "
            f"Last error: {exc2}"
        ) from exc2


# ── Quick backtest for max mode ───────────────────────────────────────────────

def _quick_backtest(manifest: AlgoManifest, bars: list[Bar]) -> BTResult:
    """Run a minimal backtest (≤_QUICK_BACKTEST_BARS bars) for scoring."""
    short_bars = bars[:_QUICK_BACKTEST_BARS]
    algo = ManifestAlgo(manifest)
    cfg = BTConfig(initial_balance=10_000.0)
    engine = BacktestEngine(algo, cfg)
    return engine.run(short_bars, symbol="EURUSD", tf=3600)


def _load_canonical_bars() -> list[Bar]:
    """Load the canonical bar fixture for quick-backtest scoring."""
    from .fixtures import load_bars
    try:
        return load_bars("bars_eurusd_h1_2y.json")
    except FileNotFoundError:
        # Return a minimal stub if fixture not available (graceful degradation)
        logger.warning("Canonical bar fixture not found; quick-backtest will score 0")
        return []


# ── Public generators ─────────────────────────────────────────────────────────

def generate_fast(
    brief: str,
    provider: LLMProvider,
    seed: int = 42,
    model: str = _DEFAULT_MODEL_FAST,
) -> tuple[AlgoManifest, GenerationTrace]:
    """Single-pass fast generation.

    Parameters
    ----------
    brief:
        Natural-language strategy brief.
    provider:
        LLM provider (local Ollama).
    seed:
        Determinism seed passed to LLM.
    model:
        Override model name; defaults to llama3.1:8b.
    """
    trace = GenerationTrace(mode="fast", model=model, seed=seed)

    system_content = FAST_SYSTEM
    user_content = render_fast_user(brief)

    messages = [
        ChatMessage(role="system", content=system_content),
        ChatMessage(role="user",   content=user_content),
    ]
    _record_turn(trace, "system", system_content)
    _record_turn(trace, "user",   user_content)

    manifest_dict, manifest = _generate_with_retry(
        provider, messages, model, seed, trace
    )

    return manifest, trace


def generate_balanced(
    brief: str,
    provider: LLMProvider,
    seed: int = 42,
    model: str = _DEFAULT_MODEL_BALANCED,
) -> tuple[AlgoManifest, GenerationTrace]:
    """Two-pass balanced generation: initial manifest + critique + revision.

    Parameters
    ----------
    brief:
        Natural-language strategy brief.
    provider:
        LLM provider.
    seed:
        Determinism seed.
    model:
        Override model name; defaults to qwen2.5:32b.
    """
    trace = GenerationTrace(mode="balanced", model=model, seed=seed)

    system_content = BALANCED_SYSTEM
    user_content = render_balanced_user(brief)

    messages: list[ChatMessage] = [
        ChatMessage(role="system", content=system_content),
        ChatMessage(role="user",   content=user_content),
    ]
    _record_turn(trace, "system", system_content)
    _record_turn(trace, "user",   user_content)

    # Turn 1: initial generation
    initial_dict, _initial_manifest = _generate_with_retry(
        provider, messages, model, seed, trace
    )

    # Build critique conversation
    critique_content = render_balanced_critique(initial_dict)
    messages = messages + [
        ChatMessage(role="assistant", content=json.dumps(initial_dict, indent=2)),
        ChatMessage(role="user",      content=critique_content),
    ]
    _record_turn(trace, "user", critique_content)

    # Turn 2: critique + revision
    revised_dict, revised_manifest = _generate_with_retry(
        provider, messages, model, seed, trace
    )

    return revised_manifest, trace


def generate_max(
    brief: str,
    provider: LLMProvider,
    seed: int = 42,
    n_candidates: int = _MAX_CANDIDATES_DEFAULT,
    reasoning_provider: Optional[LLMProvider] = None,
    model: str = _DEFAULT_MODEL_MAX,
) -> tuple[AlgoManifest, GenerationTrace]:
    """N-candidate ensemble max generation.

    Pipeline:
      1. Generate N candidates (via reasoning_provider or main provider).
      2. Quick-backtest each to score by Sharpe.
      3. Take top 2 by Sharpe.
      4. Critique each top-2 candidate, get revised manifest.
      5. Quick-backtest the 2 revised manifests.
      6. Return the best revised manifest.

    Parameters
    ----------
    brief:
        Natural-language strategy brief.
    provider:
        Main LLM provider (used for generation if reasoning_provider=None).
    seed:
        Determinism seed.
    n_candidates:
        Number of initial candidates to generate (default 5).
    reasoning_provider:
        Optional separate reasoning-model provider. If None, falls back to
        main provider with a warning recorded in the trace.
    model:
        Model name for the reasoning provider.
    """
    # Decide which provider to use for generation
    gen_provider: LLMProvider
    if reasoning_provider is not None:
        gen_provider = reasoning_provider
        gen_model = model
    else:
        gen_provider = provider
        gen_model = model
        warnings.warn(
            "generate_max: reasoning_provider=None, falling back to main provider. "
            "Set AF_ALGO_GEN_REASONING_* env vars for true max mode.",
            stacklevel=2,
        )

    trace = GenerationTrace(
        mode="max",
        model=gen_model,
        reasoning_model=gen_model if reasoning_provider else None,
        seed=seed,
    )
    if reasoning_provider is None:
        trace.warnings.append(
            "reasoning_provider=None: fell back to main provider for generation"
        )

    bars = _load_canonical_bars()

    # Context hints to diversify candidates
    _context_hints = [
        "momentum breakout above resistance using ATR bands",
        "mean reversion to moving average after oversold RSI signal",
        "volatility contraction then expansion with Bollinger bands",
        "trend following with ADX filter for strong directional markets",
        "multi-timeframe confluence using EMA slopes and RSI divergence",
        "stochastic oscillator crossover with volume confirmation",
        "keltner channel breakout with OBV divergence filter",
    ]

    # ── Phase 1: Generate N candidates ────────────────────────────────────────
    system_content = MAX_SYSTEM
    _record_turn(trace, "system", system_content)

    candidates: list[CandidateSummary] = []

    for i in range(n_candidates):
        hint = _context_hints[i % len(_context_hints)]
        user_content = render_max_user(
            brief=brief,
            candidate_num=i + 1,
            n_candidates=n_candidates,
            context_hint=hint,
        )
        messages = [
            ChatMessage(role="system", content=system_content),
            ChatMessage(role="user",   content=user_content),
        ]
        _record_turn(trace, "user", user_content)

        summary = CandidateSummary(index=i, manifest=None, sharpe=None, trades=None, max_dd=None)

        try:
            manifest_dict, manifest = _generate_with_retry(
                gen_provider, messages, gen_model, seed + i, trace
            )
            summary.manifest = manifest_dict

            # Quick backtest
            if bars:
                try:
                    result = _quick_backtest(manifest, bars)
                    summary.sharpe = result.sharpe
                    summary.trades = result.total_trades
                    summary.max_dd = result.max_drawdown
                except Exception as bt_err:
                    summary.error  = f"backtest: {bt_err}"
                    summary.sharpe = -999.0
                    summary.trades = 0
                    summary.max_dd = 100.0
            else:
                summary.sharpe = 0.0
                summary.trades = 0
                summary.max_dd = 100.0

        except GenerationError as gen_err:
            summary.error  = str(gen_err)
            summary.sharpe = -999.0

        candidates.append(summary)

    trace.candidates = candidates

    # ── Phase 2: Pick top 2 by Sharpe ─────────────────────────────────────────
    valid = [c for c in candidates if c.manifest is not None]
    if not valid:
        raise GenerationError("All candidates failed to generate valid manifests")

    valid.sort(key=lambda c: (c.sharpe or -999.0), reverse=True)
    top2 = valid[:2]

    # ── Phase 3: Critique and revise top 2 ────────────────────────────────────
    refined: list[tuple[AlgoManifest, float]] = []

    for cand in top2:
        cand.critiqued = True
        critique_content = render_max_critique(
            manifest_dict=cand.manifest or {},
            trades=cand.trades or 0,
            sharpe=cand.sharpe or 0.0,
            max_dd=cand.max_dd or 100.0,
        )

        messages = [
            ChatMessage(role="system", content=MAX_SYSTEM),
            ChatMessage(role="user",   content=critique_content),
        ]
        _record_turn(trace, "user", critique_content)

        pair = CritiquePair(
            candidate_index=cand.index,
            critique_prompt=critique_content,
            revised_manifest=None,
        )

        try:
            rev_dict, rev_manifest = _generate_with_retry(
                gen_provider, messages, gen_model, seed + 100 + cand.index, trace
            )
            pair.revised_manifest = rev_dict

            # Score revised
            if bars:
                try:
                    rev_result = _quick_backtest(rev_manifest, bars)
                    rev_sharpe = rev_result.sharpe
                except Exception:
                    rev_sharpe = cand.sharpe or 0.0
            else:
                rev_sharpe = cand.sharpe or 0.0

            refined.append((rev_manifest, rev_sharpe))

        except GenerationError as err:
            pair.error = str(err)
            # Fall back to original candidate
            try:
                original_manifest = validate_manifest(cand.manifest or {})
                refined.append((original_manifest, cand.sharpe or 0.0))
            except Exception:
                pass

        trace.critique_pairs.append(pair)

    if not refined:
        # Final fallback: use best original candidate
        best_cand = valid[0]
        best_manifest = validate_manifest(best_cand.manifest or {})
        best_cand.final = True
        return best_manifest, trace

    # Pick winner by Sharpe
    refined.sort(key=lambda x: x[1], reverse=True)
    winner_manifest, _ = refined[0]

    # Mark the winning candidate
    for cand in candidates:
        if (
            cand.manifest is not None
            and cand.manifest.get("name") == winner_manifest.name
        ):
            cand.final = True
            break

    return winner_manifest, trace

File: cpp/src/algo_gen/generator.cpp
/**
 * AlgoForge — src/algo_gen/generator.cpp
 *
 * LLM-driven manifest generation (S1) — generate_fast.
 * Python oracle: python/algoforge/algo_gen/generator.py
 *
 * Sandbox/escape-hatch is intentionally NOT here: it is a Python-subprocess
 * executor with no honest C++ analog (the engine skips escape-hatch manifests
 * in validate()). Balanced/Max modes are follow-on increments.
 */
#include "algo_gen_internal.hpp"
#include "core/algo_gen.hpp"

#include <algorithm>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

namespace algoforge::algo_gen {
namespace generator {

using algoforge::llm::ChatMessage;
using algoforge::llm::ChatRequest;
using algoforge::llm::ChatResponse;
using algoforge::llm::LLMProvider;

namespace {

const char* FIX_JSON_PROMPT =
    "The JSON you produced is malformed or schema-invalid. "
    "Please output ONLY the corrected manifest JSON in a ```json ... ``` fence. "
    "No prose, no explanation — just the fixed JSON.";

void emit(const json::JsonValue& v, std::string& out) {
    using JT = json::JsonType;
    switch (v.type) {
        case JT::Null:   out += "null"; break;
        case JT::Bool:   out += json::emit_bool(v.boolean); break;
        case JT::Number: out += json::emit_double(v.num); break;
        case JT::String: out += json::emit_string_val(v.str); break;
        case JT::Array:  out += "["; for (size_t i = 0; i < v.arr.size(); ++i) { if (i) out += ","; emit(v.arr[i], out); } out += "]"; break;
        case JT::Object: {
            out += "{"; bool f = true;
            for (const auto& [k, val] : v.obj) { if (!f) out += ","; f = false; out += json::emit_string_val(k); out += ":"; emit(val, out); }
            out += "}"; break;
        }
    }
}
std::string dumps(const json::JsonValue& v) { std::string s; emit(v, s); return s; }

std::string chat(LLMProvider& p, const std::vector<ChatMessage>& msgs, const std::string& model, int seed) {
    ChatRequest req;
    req.model = model; req.messages = msgs; req.temperature = 0.5f;
    req.max_tokens = 2048; req.seed = seed;
    ChatResponse resp = p.chat(req);
    return resp.message.content;
}

/* extract JSON + schema-validate; throws std::exception on failure. */
std::pair<json::JsonValue, AlgoManifest> extract_validate(const std::string& raw) {
    json::JsonValue d = prompts::extract_json_block(raw);  /* throws if no JSON */
    AlgoManifest m = parse_manifest(dumps(d));             /* throws on schema error */
    return {d, m};
}

/* One LLM call + parse, with a single fix-JSON retry. Records turns +
 * parse_failures into trace. Throws GenerationError after two failures. */
std::pair<json::JsonValue, AlgoManifest> generate_with_retry(
        LLMProvider& provider, std::vector<ChatMessage> messages,
        const std::string& model, int seed, GenerationTrace& trace) {
    std::string raw = chat(provider, messages, model, seed);
    trace.turns.push_back({"assistant", raw});
    try { return extract_validate(raw); }
    catch (const std::exception& e) { trace.parse_failures.push_back({raw, e.what()}); }

    std::vector<ChatMessage> retry = messages;
    retry.push_back({"assistant", raw});
    retry.push_back({"user", FIX_JSON_PROMPT});
    trace.turns.push_back({"user", FIX_JSON_PROMPT});

    std::string raw2 = chat(provider, retry, model, seed);
    trace.turns.push_back({"assistant", raw2});
    try { return extract_validate(raw2); }
    catch (const std::exception& e2) {
        trace.parse_failures.push_back({raw2, e2.what()});
        throw GenerationError(std::string("Failed to produce a valid manifest after 2 attempts. Last error: ") + e2.what());
    }
}

} /* anonymous namespace */

GenResult generate_fast(const std::string& brief, LLMProvider& provider,
                        int seed, const std::string& model) {
    GenerationTrace trace;
    trace.mode = "fast"; trace.model = model; trace.seed = seed;

    std::string sys = prompts::FAST_SYSTEM;
    std::string usr = prompts::render_fast_user(brief);
    std::vector<ChatMessage> messages = {{"system", sys}, {"user", usr}};
    trace.turns.push_back({"system", sys});
    trace.turns.push_back({"user",   usr});

    auto [d, m] = generate_with_retry(provider, messages, model, seed, trace);
    (void)d;
    return {m, trace};
}

GenResult generate_balanced(const std::string& brief, LLMProvider& provider,
                            int seed, const std::string& model) {
    GenerationTrace trace;
    trace.mode = "balanced"; trace.model = model; trace.seed = seed;

    std::string sys = prompts::BALANCED_SYSTEM;
    std::string usr = prompts::render_balanced_user(brief);
    std::vector<ChatMessage> messages = {{"system", sys}, {"user", usr}};
    trace.turns.push_back({"system", sys});
    trace.turns.push_back({"user",   usr});

    /* Turn 1: initial manifest */
    auto [initial_d, initial_m] = generate_with_retry(provider, messages, model, seed, trace);
    (void)initial_m;

    /* Turn 2: critique + revision */
    std::string critique = prompts::render_balanced_critique(initial_d);
    messages.push_back({"assistant", dumps(initial_d)});
    messages.push_back({"user", critique});
    trace.turns.push_back({"user", critique});

    auto [revised_d, revised_m] = generate_with_retry(provider, messages, model, seed, trace);
    (void)revised_d;
    return {revised_m, trace};
}

GenResult generate_max(const std::string& brief, LLMProvider& provider, int seed,
                       int n_candidates, LLMProvider* reasoning_provider,
                       const std::string& model) {
    LLMProvider& gen = reasoning_provider ? *reasoning_provider : provider;

    GenerationTrace trace;
    trace.mode = "max"; trace.model = model; trace.seed = seed;
    trace.reasoning_model = reasoning_provider ? model : std::string();
    if (!reasoning_provider)
        trace.warnings.push_back("reasoning_provider=None: fell back to main provider for generation");

    static const char* HINTS[] = {
        "momentum breakout above resistance using ATR bands",
        "mean reversion to moving average after oversold RSI signal",
        "volatility contraction then expansion with Bollinger bands",
        "trend following with ADX filter for strong directional markets",
        "multi-timeframe confluence using EMA slopes and RSI divergence",
        "stochastic oscillator crossover with volume confirmation",
        "keltner channel breakout with OBV divergence filter",
    };
    const int N_HINTS = 7;

    const std::string sys = prompts::MAX_SYSTEM;
    trace.turns.push_back({"system", sys});

    struct Cand { CandidateSummary s; AlgoManifest m; bool ok = false; };
    std::vector<Cand> cands;

    /* Phase 1: generate n candidates (no canonical bars → graceful sharpe=0). */
    for (int i = 0; i < n_candidates; ++i) {
        std::string usr = prompts::render_max_user(brief, i + 1, n_candidates, HINTS[i % N_HINTS]);
        std::vector<ChatMessage> messages = {{"system", sys}, {"user", usr}};
        trace.turns.push_back({"user", usr});

        Cand c; c.s.index = i; c.s.max_dd = 100.0;
        try {
            auto [d, m] = generate_with_retry(gen, messages, model, seed + i, trace);
            (void)d;
            c.m = m; c.ok = true; c.s.has_manifest = true; c.s.name = m.name; c.s.sharpe = 0.0;
        } catch (const GenerationError& e) {
            c.s.error = e.what(); c.s.sharpe = -999.0;
        }
        cands.push_back(std::move(c));
    }
    for (const auto& c : cands) trace.candidates.push_back(c.s);

    /* Phase 2: top-2 valid candidates by sharpe. */
    std::vector<int> valid;
    for (int i = 0; i < (int)cands.size(); ++i) if (cands[i].ok) valid.push_back(i);
    if (valid.empty()) throw GenerationError("All candidates failed to generate valid manifests");
    std::sort(valid.begin(), valid.end(), [&](int a, int b){ return cands[a].s.sharpe > cands[b].s.sharpe; });
    std::vector<int> top2(valid.begin(), valid.begin() + std::min<size_t>(2, valid.size()));

    /* Phase 3: critique + revise the top 2. */
    std::vector<std::pair<AlgoManifest,double>> refined;
    for (int idx : top2) {
        trace.candidates[idx].critiqued = true;
        std::string crit = prompts::render_max_critique(json::JsonValue{}, cands[idx].s.trades,
                                                        cands[idx].s.sharpe, cands[idx].s.max_dd);
        std::vector<ChatMessage> messages = {{"system", sys}, {"user", crit}};
        trace.turns.push_back({"user", crit});
        try {
            auto [rd, rm] = generate_with_retry(gen, messages, model, seed + 100 + cands[idx].s.index, trace);
            (void)rd;
            refined.push_back({rm, cands[idx].s.sharpe});
        } catch (const GenerationError&) {
            refined.push_back({cands[idx].m, cands[idx].s.sharpe});  /* fall back to original */
        }
    }
    if (refined.empty()) return {cands[valid[0]].m, trace};

    std::sort(refined.begin(), refined.end(), [](const auto& a, const auto& b){ return a.second > b.second; });
    AlgoManifest winner = refined[0].first;
    for (auto& c : trace.candidates) if (c.has_manifest && c.name == winner.name) { c.is_final = true; break; }
    return {winner, trace};
}

} /* namespace generator */
} /* namespace algoforge::algo_gen */
